src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,462 @@
+                                       (!string.IsNullOrEmpty(previousValue.OutputPath) && !string.Equals(previousValue.OutputPath, value.OutputPath, StringComparisons.Paths));
+
+            // Project properties and shared items cause all TempPEs to be dirty, because shared items are included in all TempPEs
+            if (hasHadProjectChange)
+            {
+                foreach (string item in newDesignTimeInputs)
+                {
+                    await FireTempPEDirtyAsync(item, value.ShouldCompile);
+                }
+            }
+            else
+            {
+                // otherwise just the ones that we've added are dirty
+                foreach (string item in addedDesignTimeInputs)
+                {
+                    await FireTempPEDirtyAsync(item, value.ShouldCompile);
+                }
+            }
+            // We don't recompile for shared design time inputs
+            if (hasRemovedDesignTimeSharedInputs || hasAddedDesignTimeSharedInputs)
+            {
+                // adding or removing shared design time inputs dirties things but doesn't recompile
+                foreach (string item in newDesignTimeInputs)
+                {
+                    // We don't want to fire again if we already fired above and compiled
+                    if (!addedDesignTimeInputs.Contains(item))
+                    {
+                        await FireTempPEDirtyAsync(item, false);
+                    }
+                }
+            }
+            foreach (string item in removedDesignTimeInputs)
+            {
+                _buildManager.OnDesignTimeOutputDeleted(item);
+            }
+        }
+
+        /// <summary>
+        /// InitializeInnerCoreAsync is responsible for setting an initial AppliedValue. This value will be used by any UI thread calls that may happen
+        /// before the first data flow blocks have been processed. If this method doesn't return a value then the system will block until the first blocks
+        /// have been applied but since we are initialized on the UI thread, that blocks for us, so we must return an initial state.
+        /// </summary>
+        protected override Task InitializeInnerCoreAsync(CancellationToken cancellationToken)
+        {
+            AppliedValue = new ProjectVersionedValue<DesignTimeInputsItem>(new DesignTimeInputsItem(), ImmutableDictionary.Create<NamedIdentity, IComparable>());
+
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// This method is where we tell data flow which blocks we're interested in receiving updates for
+        /// </summary>
+        protected override IDisposable LinkExternalInput(ITargetBlock<IProjectVersionedValue<InputTuple>> targetBlock)
+        {
+            return ProjectDataSources.SyncLinkTo(
+                _projectSubscriptionService.ProjectSource.SourceBlock.SyncLinkOptions(),
```

Comment:
Below when I said don't read from "ProjectInstance", You want ProjectRuleSource passing it ConfigurationGeneral similar to generators: http://index/?leftProject=Microsoft.VisualStudio.ProjectSystem&leftSymbol=ixccm82ohk0j&rightProject=Microsoft.VisualStudio.ProjectSystem.VS.Implementation&file=CodeGenerators%5CSingleFileGeneratorsService.cs&line=158.

Reply:
Thank you, I never did see ProjectRuleSource ever have any information in it, and the extra things passed to SyncLinkOptions probably explains that.